support `funcade` to refresh tokens on demand

VigneshwaranJheyaraman opened the pull request:
## preface
With `funcade` middleware to fetch `jwks-keyset` while starting `reitit-middleware` + `buddy-middleware` for certain `authenticator` it may require to refresh the token periodically to obtain new `kid` for `jwks` authentication. In order to make sure authentication still happens we need to refresh those `keyset` and for which we are exposing functions which actually refreshes the `keyset` and updates the refrence for the `variable`

 ## fixes done
* Added `jwks->keys-fn` which returns the `fn` which returns the `cert` given the `kid`
* Added `refresh-keyset` which actually refreshes the `keyset` and updates the `variable-reference`
* Made the `current-kids` _dynamic_
* Updated `buddy-middleware` which supports `refresh-keyset?` option which actually enables `refresh` support via `clojure-fn`
* Added `find-kid` function which extracts `kid` from `token`
* Modified `find-token-key` to make use of `find-kid` function
* Changes are working fine validated locally

 ## unit testing
```clojure
dev=> funcade.jwks/current-kids
("kid1", "kid2", "kid3")
dev=> (funcade.jwks/refresh-kids)
"[funcade]: done refreshing keyset"
nil
dev=> funcade.jwks/current-kids
("kid1.1", "kid2.2", "kid3.3")
```

Commit:
support `funcade` to refresh tokens on demand

 ## preface
With `funcade` middleware to fetch `jwks-keyset` while starting `reitit-middleware` + `buddy-middleware` for certain `authenticator` it
may require to refresh the token periodically to obtain new `kid` for `jwks` authentication. In order to make sure authentication still
happens we need to refresh those `keyset` and for which we are exposing functions which actually refreshes the `keyset` and updates the
refrence for the `variable`

 ## fixes done
* Added `jwks->keys-fn` which returns the `fn` which returns the `cert` given the `kid`
* Added `refresh-keyset` which actually refreshes the `keyset` and
  updates the `variable-reference`
* Made the `current-kids` _dynamic_
* Updated `buddy-middleware` which supports `refresh-keyset?` option
  which actually enables `refresh` support via `clojure-fn`
* Added `find-kid` function which extracts `kid` from `token`
* Modified `find-token-key` to make use of `find-kid` function
* Changes are working fine validated locally

 ## unit testing
```clojure
dev=> funcade.jwks/current-kids
("kid1", "kid2", "kid3")
dev=> (funcade.jwks/refresh-kids)
"[funcade]: done refreshing keyset"
nil
dev=> funcade.jwks/current-kids
("kid1.1", "kid2.2", "kid3.3")
```

Commit:
@tolitius suggestion

 ## fixes done
* Removing all the logic to make use of `alter-var-root` which kind of
  complicates stuff. Now using `private-atom` which is visible only
  inside `#'funcade.jwks` namespace and exposing `setters` and `getters` as
  `functions` which relevant `args`

tolitius commented on project.clj:
```diff
@@ -1,4 +1,4 @@
-(defproject tolitius/funcade "0.1.24-SNAPSHOT"
+(defproject tolitius/funcade "0.1.25"
```
`0.1.24`

tolitius replied:
👆

tolitius commented on src/funcade/jwks.clj:
```diff
@@ -2,23 +2,38 @@
+(defn- find-key-by-kid
+  "Given the token's kid find the key from keyset"
+  [& kids]
+  (when (and (some? @keyset)
+           (seq kids))
```
format:
```clojure
  (when (and (some? @keyset)
             (seq kids))
```

tolitius commented on src/funcade/jwks.clj:
```diff
@@ -2,23 +2,38 @@
-  keyset (atom {}))
+  keyset (atom nil))
+
+(defonce ^:private
+  uri (atom nil))
```
why do we need to keep a changeable state for the URI?
`uri` should come from config, as a function argument

if there is no way we can keep it clean: function arg, we should keep it in "`keyset`" atom: `{:keys ... :uri ...}`

tolitius commented on src/funcade/jwks.clj:
```diff
@@ -32,9 +47,31 @@
-(defn find-token-key [jwks token]
+(defn refresh-kids
+  "Refresh new set of jwks-keyset from auth-provider"
+  []
```
this should _take_ and `uri`

the idea is to minimize the state handling

tolitius commented on src/funcade/middleware/buddy.clj:
```diff
@@ -61,5 +64,7 @@
    i.e. needs a 'https://foo.com/bar/jwks' URI that returns unsign keys"
-  [{:keys [uri] :as options}]
-  (jwks-backend (assoc options :keyset (jk/jwks->keys uri))))
+  [{:keys [uri refresh-keyset?] :as options}]
+  (jwks-backend (cond-> options
+                       (not refresh-keyset?)  (assoc :keyset (jwks/jwks->keys uri))
+                       refresh-keyset?        (assoc :keyset-fn (jwks/jwks->keys-fn uri)))))
```
refreshing keyset should be an internal ability of funcade
it needs to take an interval (i.e. number of seconds)
and on every interval it just `swap!`s the atom with a new values

Commit:
Supporting `refresh` functionality within `funcade`

 ## Fixes done
* On discussing with @tolitius we are including support for
`scheduled-task` which refreshed the token given the `interval-ms` and
also supports `refresh-callback` which triggers for every `refresh` task
triggers
* Updated README

tolitius commented on README.md:
````diff
@@ -142,6 +143,71 @@ invalid scope:
 ;; {:status 401, :body {:message "missing required scope", :required :my-scope, :scopes (:not-my-scope)}}
 ```
 
+### Support middleware `keyset` refresh
````
"Support middleware `keyset` refresh" => "`keyset` refresh"
it already lives under the "`use in middleware`" section

tolitius commented on src/funcade/jwks.clj:
```diff
@@ -2,39 +2,77 @@
+(defn find-current-kids
+  "Find all current kids"
+  []
+  (when-let [keyset-value @keyset]
+    (keys keyset-value)))
```
```clojure
  (some-> @keyset keys)
```

tolitius commented on src/funcade/jwks.clj:
```diff
@@ -2,39 +2,77 @@
           throw))))
 
-(defn find-token-key [jwks token]
+(defn refresh-kids
+  "Refresh new set of jwks-keyset from auth-provider"
+  [uri callback]
+  (let [refreshed-keyset (jwks->keys uri)]
+    (prn "[funcade]: refreshed keyset")
+    (callback (keys refreshed-keyset))
+    (keys refreshed-keyset)))
```
```clojure
(defn refresh-kids
  "refresh jwks keyset from auth provider"
  [uri callback]
  (let [keyset (-> uri jwks->keys keys)
        fun (or callback
                (fn [] (prn "[funcade]: refreshed keyset")))]

    (fun keyset)
    keyset))
```

tolitius commented on src/funcade/jwks.clj:
```diff
@@ -2,39 +2,77 @@
+
+(defn- find-token-key-by-kid
+  "Given the token's kid find the key from keyset"
+  [& kids]
```
why `&`?
if the function _expects_ kids, it should not be optional

tolitius commented on src/funcade/jwks.clj:
```diff
@@ -2,39 +2,77 @@
+        :or {refresh-callback identity}}]
+  (let [_ (jwks->keys url)]
+    (when (some? refresh-interval-ms)
+      (scheduler/every refresh-interval-ms
+                       (partial refresh-kids url refresh-callback)))
```
it should return this scheduler
so an "`(scheduler/stop ..)`" can be called on "destruction"

otherwise it is going to be scheduled many times over the dev session
[as shown in docs](https://github.com/tolitius/yang?tab=readme-ov-file#schedule)

tolitius commented on src/funcade/jwks.clj:
```diff
@@ -2,39 +2,77 @@
+  scheduler to refresh keyset for every interval"
+  [url {:keys [refresh-interval-ms
+               refresh-callback]
+        :or {refresh-callback identity}}]
```
don't default to identity, so 
```clj
fun (or callback
        (fn [] (prn "[funcade]: refreshed keyset")))
```
from the above comment works

OR default it to `(prn ...)` here if not provided right away

tolitius commented on src/funcade/jwks.clj:
```diff
@@ -2,39 +2,77 @@
+    (callback (keys refreshed-keyset))
+    (keys refreshed-keyset)))
+
+(defn jwks->keys-fn
```
it should not couple:
* getting the keyset
* setting the initial state
* sheduling the refresh

and
* returning a function it does not create (or knows nothing about)

this function should be called "`jwks->keyset`", and it should do:
* `jwks->keys`
* schedule

that's it

tolitius commented on src/funcade/middleware/buddy.clj:
```diff
@@ -61,5 +61,7 @@
+  [{:keys [uri]
+    :as options}]
+  (jwks-backend (assoc options
+                       :keyset-fn (jwks/jwks->keys-fn uri options))))
```
```clojure
:keyset-fn (jwks/jwks->keys-fn uri options)
```
we don't need it

tolitius commented on src/funcade/middleware/buddy.clj:
```diff
@@ -28,7 +28,7 @@
 
 (defn jwks-backend
-  [{:keys [keyset authfn unauthorized-handler options token-name on-error]
+  [{:keys [_keyset authfn unauthorized-handler options token-name on-error keyset-fn]
```
should not take "`keyset-fn`"

instead should call "`jwks->keyset`"

tolitius commented on src/funcade/middleware/buddy.clj:
```diff
@@ -39,7 +39,7 @@
     (-authenticate [_ request data]
       (try
-        (let [tkey (jk/find-token-key keyset data)]
+        (let [tkey (keyset-fn (jwks/find-kid data))]
```
`keyset-fn` => `find-token-key-by-kid`

Commit:
@tolitius changes requested + handle `singleton` scheduler

 ## preface
@tolitius added some comments, which are fixed and alos taking a point

> _`funcade` has to handle the scheduling within library_

For the very same imported `tolitius/yang` library which helps with
`yang.scheduler` to run a scheduled job. Also since `funcade` handles
`reitit` / `buddy` middleware it might be attached with a
`ring-handler`. The **problem** we might have is since its `handler-fn`
gets associated with `middleware`, when it gets `attached` to
`http-server` the `handler` can be shared across to different
`thread-workers` resulting in multiple scheduler. So made sure that
only have **one scheduler**

 ## fixes done
* Updated readme
* replacing `if-let` / `when-let` with `some->`
* made sure of `scheduler` to be initialized only once and not any more

tolitius commented on README.md:
```diff
@@ -10,6 +10,7 @@ creates, manages and refreshes jwt tokens
   - [group many sources](#group-many-sources)
 - [use key sets (JWKS)](#use-key-sets-jwks)
 - [use in middleware](#using-middleware)
+  - [support JWKS keyset refresh](#support-keyset-refresh)
```
`"#support-keyset-refresh"` won't click into "`keyset` refresh section"
